src/manifold/src/impl.h

```diff
@@ -65,7 +65,183 @@ struct Manifold::Impl {
   Impl(Shape, const mat4x3 = mat4x3(1));
 
-  Impl(const MeshGL&, std::vector<float> propertyTolerance = {});
+  template <typename Precision>
+  Impl(const MeshGLP<Precision>& meshGL,
+       std::vector<Precision> propertyTolerance) {
+    const auto numVert = meshGL.NumVert();
+    const auto numTri = meshGL.NumTri();
+
+    if (meshGL.numProp < 3) {
+      MarkFailure(Error::MissingPositionProperties);
+      return;
+    }
+
+    if (meshGL.mergeFromVert.size() != meshGL.mergeToVert.size()) {
+      MarkFailure(Error::MergeVectorsDifferentLengths);
+      return;
+    }
+
+    if (!meshGL.runTransform.empty() &&
+        12 * meshGL.runOriginalID.size() != meshGL.runTransform.size()) {
+      MarkFailure(Error::TransformWrongLength);
+      return;
+    }
+
+    if (!meshGL.runOriginalID.empty() && !meshGL.runIndex.empty() &&
+        meshGL.runOriginalID.size() + 1 != meshGL.runIndex.size() &&
+        meshGL.runOriginalID.size() != meshGL.runIndex.size()) {
+      MarkFailure(Error::RunIndexWrongLength);
+      return;
+    }
+
+    if (!meshGL.faceID.empty() && meshGL.faceID.size() != meshGL.NumTri()) {
+      MarkFailure(Error::FaceIDWrongLength);
+      return;
+    }
+
+    std::vector<int> prop2vert(numVert);
+    std::iota(prop2vert.begin(), prop2vert.end(), 0);
+    for (size_t i = 0; i < meshGL.mergeFromVert.size(); ++i) {
+      const int from = meshGL.mergeFromVert[i];
+      const int to = meshGL.mergeToVert[i];
+      if (from >= static_cast<int>(numVert) ||
+          to >= static_cast<int>(numVert)) {
+        MarkFailure(Error::MergeIndexOutOfBounds);
+        return;
+      }
+      prop2vert[from] = to;
+    }
+
+    Vec<TriRef> triRef;
+    Vec<ivec3> triProperties;
+    if (meshGL.numProp > 3) {
+      triProperties.resize(numTri);
+      for (size_t i = 0; i < numTri; ++i) {
+        for (const size_t j : {0, 1, 2}) {
+          triProperties[i][j] = meshGL.triVerts[3 * i + j];
+        }
+      }
+    }
+
+    const auto numProp = meshGL.numProp - 3;
+    meshRelation_.numProp = numProp;
+    meshRelation_.properties.resize(meshGL.NumVert() * numProp);
+    // This will have unreferenced duplicate positions that will be removed by
+    // Impl::RemoveUnreferencedVerts().
+    vertPos_.resize(meshGL.NumVert());
+
+    for (size_t i = 0; i < meshGL.NumVert(); ++i) {
+      for (const int j : {0, 1, 2})
+        vertPos_[i][j] = meshGL.vertProperties[meshGL.numProp * i + j];
+      for (size_t j = 0; j < numProp; ++j)
+        meshRelation_.properties[i * numProp + j] =
+            meshGL.vertProperties[meshGL.numProp * i + 3 + j];
+    }
+
+    halfedgeTangent_.resize(meshGL.halfedgeTangent.size() / 4);
+    for (size_t i = 0; i < halfedgeTangent_.size(); ++i) {
+      for (const int j : {0, 1, 2, 3})
+        halfedgeTangent_[i][j] = meshGL.halfedgeTangent[4 * i + j];
+    }
+
+    if (meshGL.runOriginalID.empty()) {
+      // FIXME: should we do this?
+      meshRelation_.originalID = Impl::ReserveIDs(1);
```

Comment:
Huh, it does seem to conflict with the statement at the end of the function. Does anything break if you remove it? I should probably think through the cases here...